### PR TITLE
Make NCE_OP more efficient and support SelectedRows

### PR DESCRIPTION
**Pull request by seiriosPlus**

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_nce.py [hunk 1]
```diff
@@ -107,10 +108,15 @@ def test_check_grad(self):
...
-class TestNCECase1(TestNCE):
+class TestNCECase1Tensor(TestNCE):
     def set_data(self):
-        self.generate_data(10, 20, 10, 2, 5)
+        self.generate_data(10, 20, 10, 2, 5, False)
 
 
+# class TestNCECase1SelectedRows(TestNCE):
+#     def set_data(self):
+#         self.generate_data(10, 20, 10, 2, 5, True)
```
Comment (panyx0718) [inline python/paddle/fluid/tests/unittests/test_nce.py]: why commented?

Comment (seiriosPlus) [reply python/paddle/fluid/tests/unittests/test_nce.py]: I have added a new one.

**Comment by seiriosPlus** [comment]
Run training with ```https://github.com/jacquesqiao/models/blob/add-word2vec/fluid/PaddleRec/word2vec/train.py```

SGD + BATCH_SIZE=100 + 1000 BATCH + SPARSE , Cost:  21.358247 Sec
SGD + BATCH_SIZE=100 + 1000 BATCH + DENSE ,   Cost:  251.58735 Sec

**Review by wanghaoshuang — DISMISSED** [review]
LGTM

**Comment by jacquesqiao** [comment]
pull一下develop最新代码，这个里边有很多已经merge了

File: paddle/fluid/operators/nce_op.h [hunk 2]
```diff
@@ -70,13 +76,13 @@ void PrepareSamples(const framework::ExecutionContext& context,
...
+  void Compute(const framework::ExecutionContext &context) const override {
     int sampler_type = context.Attr<int>("sampler");
     int seed = context.Attr<int>("seed");
     int num_total_classes = context.Attr<int>("num_total_classes");
     int num_neg_samples = context.Attr<int>("num_neg_samples");
 
-    Sampler* sampler;
+    Sampler *sampler;
     switch (sampler_type) {
       case 0: {
```
Comment (jacquesqiao) [inline paddle/fluid/operators/nce_op.h:87]: 建议定义成一个有意义的变量名，1，2，3不太容易看

**Review by typhoonzero — DISMISSED** [review]
LGAPI

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -5756,20 +5801,20 @@ def image_resize(input,
                          Default: None
         name(str|None): A name for this layer(optional). If set None, the layer
                         will be named automatically.
-        resample(str): The resample method. It supports 'BILINEAR' and 'NEAREST'
+        resample(str): The resample method. It supports 'BILINEAR' and 'NEAREST'
```
Comment (jacquesqiao) [inline python/paddle/fluid/layers/nn.py]: revert the unused code format

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -6446,15 +6481,15 @@ def affine_grid(theta, out_shape, name=None):
                         [x_14, x_15, x_16]]
                        [[x_21, x_22, x_23]
                         [x_24, x_25, x_26]]]
-
+
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py]: Are you sure your pylint is working correctly?

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -4342,11 +4343,12 @@ def nce(input,
         sampler (str): The sampler used to sample class from negtive classes.
                        It can be 'uniform', 'log_uniform' or 'custom_dist'.
                        default: 'uniform'.
-        custom_dist (Variable): A tensor with shape [num_total_classes].
+        custom_dist (float[]): A float[] with size=num_total_classes.
                        It is used when sampler is set to 'custom_dist'.
                        custom_dist[i] is the probsbility of i-th class to be sampled.
                        default: None.
         seed (int): The seed used in sampler. default: 0.
+        is_sparse(bool): The flag indicating whether to use sparse update.
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py]: Can you give more detail? What is going to be sparse?

**Review by panyx0718 — COMMENTED** [review]
For changes not related to the title of this PR, can you put them in a different PR?

File: paddle/fluid/operators/math/sampler.h [hunk 6]
```diff
@@ -108,10 +113,10 @@ class CustomSampler : public Sampler {
   float Probability(int64_t value) const override;
 
  private:
-  std::shared_ptr<std::vector<float>> alias_probs_;
-  std::shared_ptr<std::vector<int64_t>> alias_;
-  std::shared_ptr<std::vector<float>> probs_;
-  std::shared_ptr<std::mt19937_64> random_engine_;
+  const float* alias_probs_;
```
Comment (jacquesqiao) [inline paddle/fluid/operators/math/sampler.h:116]: Why change smart pointer to pointer?

File: paddle/fluid/operators/nce_op.cc [hunk 7]
```diff
@@ -190,20 +213,45 @@ class NCEOpGrad : public framework::OperatorWithKernel {
...
+    auto attr = op_desc.GetAttr("is_sparse");
+    bool is_sparse = boost::get<bool>(attr);
+    if (is_sparse) {
+      VLOG(30) << "nce_op_grad op " << weight_grad << " and " << bias_grad
+               << " is set to SelectedRows";
+      block->Var(weight_grad)
+          ->SetType(framework::proto::VarType::SELECTED_ROWS);
+      block->Var(bias_grad)->SetType(framework::proto::VarType::SELECTED_ROWS);
+    } else {
+      VLOG(30) << "nce_op_grad op " << weight_grad << " and " << bias_grad
```
Comment (jacquesqiao) [inline paddle/fluid/operators/nce_op.cc:239]: is set to LoDTensor

File: python/paddle/fluid/layers/nn.py (hunk 4 above)
Comment (seiriosPlus) [reply python/paddle/fluid/layers/nn.py]: It made by my CLION, I have reverted it.

File: paddle/fluid/operators/math/sampler.h (hunk 6 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/math/sampler.h:116]: I think the smart pointer is not needed.

File: paddle/fluid/operators/nce_op.cc (hunk 7 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/nce_op.cc:239]: Done

File: python/paddle/fluid/layers/nn.py (hunk 5 above)
Comment (seiriosPlus) [reply python/paddle/fluid/layers/nn.py]: Done

**Review by typhoonzero — APPROVED** [review]
LGAPI

File: paddle/fluid/operators/math/sampler.cc [hunk 8]
```diff
@@ -60,75 +60,30 @@ float LogUniformSampler::Probability(int64_t value) const {
   return (log((value + 2.0) / (value + 1.0))) / log_range_;
 }
 
-CustomSampler::CustomSampler(int64_t range, const float* probabilities,
+CustomSampler::CustomSampler(int64_t range, const float *probabilities,
+                             const int *alias, const float *alias_probabilities,
```
Comment (panyx0718) [inline paddle/fluid/operators/math/sampler.cc:64]: This is a very strange API!

Will someone change alias_probabilities somewhere and affect the sampler?

**Review by panyx0718 — APPROVED** [review]
LG API

File: paddle/fluid/operators/math/sampler.h [hunk 9]
```diff
@@ -99,6 +103,7 @@ class LogUniformSampler : public Sampler {
 class CustomSampler : public Sampler {
  public:
   explicit CustomSampler(int64_t range, const float* probabilities,
+                         const int* alias, const float* alias_probabilities,
```
Comment (panyx0718) [inline paddle/fluid/operators/math/sampler.h:106]: Can these be changed to non-pointer?

Comment (seiriosPlus) [reply paddle/fluid/operators/math/sampler.h:106]: 我在另提一个PR修改，这个PR涉及到API比较难MERGE。

File: paddle/fluid/operators/math/sampler.cc (hunk 8 above)
Comment (seiriosPlus) [reply paddle/fluid/operators/math/sampler.cc:64]: alias_probabilities will only be used in nce_op, If someone changes alias_probabilities somewhere, the sampler will be affected.

**Pull request merged**